core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -528,7 +528,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       }
 
     // Optionally scale number of executors dynamically based on workload. Exposed for testing.
-    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false)
+    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false) &&
+      (_conf.getInt("spark.executor.instances", 0) == 0)
```

Comment:
I think this is better as `! ... _conf.contains(...)`? Although I don't think 0 executors is allowed, this logic would not work if it were.

The docs in `running-on-yarn.md` need to update to explain the new behavior a little.